CDCgov/prime-reportstream
# ErrorBoundary and ErrorPage upgrades

This PR updates `<ErrorPage>` to be more configurable and adds a custom `<ErrorBoundary>` so we have more customization over how we track errors in the long run. Ideally, I would like us to be able to transfer status codes from API calls to the `ErrorPage` component for more comprehensive error messaging, but this is an mvp-esque solution.

Test Steps:

Before anything...
1. pull `kevinhaube/error-handling-rest-hooks`
2. `yarn install` and `yarn start`
3. Log in

Test `/daily-data`:
1. F12 > Application > change the value of your `global-context-org` to a random string
2. Navigate to `/daily-data` and observe the `<OrgName>` component above the Covid-19 title. This should be an error.
3. Change your `global-context-org` back to its default (or use the admin dropdown to pick an org)
4. In the code, navigate to `resources/AuthResource.ts` and replace the `Authorization` header with a random string.
5. Refresh `/daily-data`. You should see an error `Alert` where listed results would normally be.
6. Ctrl+Z to undo the changes to the `AuthResource` file.

Test '/report-details'
1. From `/daily-data`, click a report to navigate to the `/report-details` page
2. In the url, mutate the `reportId` query to a random string and hit Enter
3. You will be shown the development-only error overlay; close that to see the `ErrorPage` that will be shown in staging.
4. Navigate back to the details page with the proper query param
5. Once again, mutate your bearer token in the code and observe the changes on refresh

Additional testing can be done by commenting out endpoints in the back-end code, re-packaging, and re-deploying it to your localhost:7071 port. This is how I tested the Facilities endpoint.

## Changes
![Screenshot from 2021-11-01 11-05-33](https://user-images.githubusercontent.com/4022304/139695217-1dc8db34-739f-4d05-abb4-c571e3e23f42.png)
![Screenshot from 2021-11-01 11-04-30](https://user-images.githubusercontent.com/4022304/139695224-5408ea71-e2a0-4120-a512-a4d9bc209b6a.png)
![Screenshot from 2021-11-01 11-13-15](https://user-images.githubusercontent.com/4022304/139695231-6fdc8037-cc71-48c7-b112-35b1e51262be.png)


## Checklist

### Testing

### Process

## Fixes
- #2368 

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #2552

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
No
- Does this PR include changes in authentication and/or authorization of existing endpoints?
No
- Does this change introduce new dependencies that need vetting?
No
- Does this change require changes to our infrastructure?
No
- Does logging contain sensitive data?
No
- Does this PR include or remove any sensitive information itself?
No

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
